plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/CollationAwareQueryBuilder.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.postgresql;
+
+import io.trino.plugin.jdbc.DefaultQueryBuilder;
+import io.trino.plugin.jdbc.JdbcClient;
+import io.trino.plugin.jdbc.JdbcColumnHandle;
+import io.trino.plugin.jdbc.JdbcJoinCondition;
+import io.trino.plugin.jdbc.JdbcTypeHandle;
+import io.trino.plugin.jdbc.QueryParameter;
+import io.trino.plugin.jdbc.WriteFunction;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.type.Type;
+
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+import static io.trino.plugin.postgresql.PostgreSqlClient.isCollatable;
+import static io.trino.plugin.postgresql.PostgreSqlSessionProperties.isEnableStringPushdownWithCollate;
+import static java.lang.String.format;
+
+public class CollationAwareQueryBuilder
+        extends DefaultQueryBuilder
+{
+    @Override
+    protected String formatJoinCondition(JdbcClient client, String leftRelationAlias, String rightRelationAlias, JdbcJoinCondition condition)
+    {
+        boolean isCollatable = Stream.of(condition.getLeftColumn(), condition.getRightColumn())
+                .anyMatch(PostgreSqlClient::isCollatable);
+
+        if (isCollatable) {
+            return format(
+                    "%s.%s COLLATE \"C\" %s %s.%s COLLATE \"C\"",
+                    leftRelationAlias,
+                    buildJoinColumn(client, condition.getLeftColumn()),
+                    condition.getOperator().getValue(),
+                    rightRelationAlias,
+                    buildJoinColumn(client, condition.getRightColumn()));
+        }
+
+        return super.formatJoinCondition(client, leftRelationAlias, rightRelationAlias, condition);
+    }
+
+    @Override
+    protected String toPredicate(JdbcClient client, ConnectorSession session, JdbcColumnHandle column, JdbcTypeHandle jdbcType, Type type, WriteFunction writeFunction, String operator, Object value, Consumer<QueryParameter> accumulator)
```

Comment:
Question: Is it possible to identify the operator being applied? e.g. For =, != or IN collation is not needed. But for >, < or BETWEEN it will be needed?

Reply:
This is regarding future direction - doesn't need to be done here.

Reply:
It's possible but I guess that you still need to apply collation while doing equal comparison.

Reply:
True if someone is using a custom collation on PostgreSQL side but with most of the collations equality and inequality don't need collation.

The reason why I mention this is that with collation specified in query PostgreSQL isn't using indexes anymore so pushdown isn't as effective. For supporting more collations we'd need the engine to understand what collation is.

Anyway, this is future work and we can discuss this separately.